### ISSUE TITLE
Delayed sequence can be used for all algorithms that take random-access range?

### ISSUE THREAD

**Issue by WhateverLiu** [cmuparlay/parlaylib#79]
Dear Parlaylib maintainers,

I have been using the library with continuing excitement. For a time I thought `delayed_sequence` can only be supplied to primitives documented in Section "Delayed sequence algorithms" <https://cmuparlay.github.io/parlaylib/algorithms/delayed.html>. 

But lately, I supplied `reduce_by_key()` (not in Section "Delayed sequence algorithm") with a `delayed_sequence` due to the massive data size, and it produced the correct result.

This makes me wonder, do all algorithms accepting random-access range will work correctly with `delayed_sequence`? By its definition it seems yes, but I just want to confirm it.

Thank you!

**Comment by gblelloch** [comment]
There is a distinction between a RAD and a BID delayed sequence, which is
described in the paper: https://dl.acm.org/doi/10.1145/3503221.3508434

Any RAD delayed sequence can be used efficienctly by any algorithm that
uses a random access range.  It looks like all delayed sequences support
op[] so I imagine it works for all delayed sequences but would not
necessarily be fast for BID delayed sequences (Daniel, is this correct?)

Basically delayed map, delayed tabulate generate RADs but delayed scan
generates a BID.

Guy
